File: Kroeg.Client/js/TemplateService.ts
import { EntityStore } from "./EntityStore";
import * as AS from "./AS";
import { Session } from './Session';
import * as twemoji from "twemoji";

export enum TemplateItemType {
    Element = 0,
    Text,
    Script
}

export class TemplateItem {
    public type: TemplateItemType;
    public data: string;
    public children: TemplateItem[];
    public arguments: {[name: string]: TemplateItem[]};
    public builder: { (regs: Registers): any };
}

export class TemplateService {
    private _templatePromise: Promise<{[item: string]: TemplateItem}>

    constructor() {
        this._templatePromise = this._getTemplates();
    }

    private async _getTemplates(): Promise<{[item: string]: TemplateItem}> {
        const result = await fetch("/settings/templates");
        return await result.json();
    }

    public getTemplates(): Promise<{[item: string]: TemplateItem}> {
        return this._templatePromise;
    }
}

export class RenderResult {
    public result: HTMLElement;
    public subRender: {id: string, into: HTMLElement, template: TemplateItem, data: {[name: string]: string}, parent: AS.ASObject}[];
    public componentHandles: HTMLElement[];
}

class _ASHandler {
    constructor(private _regs: Registers) {}

    public get(name: string): any[] {
        return AS.get(this._regs.object, name);
    }

    public take(name: string, def?: any) {
        if (def === undefined)
            def = "";
        if (!AS.has(this._regs.object, name)) return def;
        let item = this.get(name)[0];
        if (item === null) return def;
        return item;
    }

    public has(name: string) {
        return AS.has(this._regs.object, name);
    }

    public contains(name: string, val: any) {
        return AS.contains(this._regs.object, name, val);
    }

    public containsAny(name: string, val: any[]) {
        return AS.containsAny(this._regs.object, name, val);
    }
}

class RendererInfo {
    constructor(private session: Session) {}

    public get client() { return true; }
    public get server() { return false; }

    public get loggedInAs() { return this.session.user; }

    private static _allowed_attributes = /^(href|rel|class)$/;
    private static _allowed_classes = /^((h|p|u|dt|e)-.*|mention|hashtag|ellipsis|invisible)$/;
    private static _disallowed_nodes = /^(script|object|embed)$/;

    public sanitize(data: string) {
        return twemoji.parse(data.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;'));
    }

    public date(data: string) {
        return (new Date(data)).toLocaleString();
    }

    private _clean(data: HTMLElement) {
        if (data.nodeType == document.TEXT_NODE) return;
        if (data.nodeType != document.ELEMENT_NODE) return;
        if (RendererInfo._disallowed_nodes.exec(data.nodeName)) {
            data.remove();
            return;
        }
        for (let attribute of Array.from(data.attributes)) {
            if (!RendererInfo._allowed_attributes.exec(attribute.name)) data.attributes.removeNamedItem(attribute.name);
        }

        for (let cl of Array.from(data.classList)) {
            if (!RendererInfo._allowed_classes.exec(cl)) data.classList.remove(cl);
        }

        for (let child of Array.from(data.childNodes)) {
            this._clean(child as HTMLElement);
        }
    }

    public clean(data: string) {
        let doc = document.createElement("div");
        doc.insertAdjacentHTML('beforeend', data);
        this._clean(doc);
        twemoji.parse(doc);
        return doc.innerHTML;
    }
}

class Registers {
    public AS: _ASHandler;
    public Renderer: RendererInfo;
    public object: AS.ASObject;
    public item: any;
    public depth: number;
    public Data: {[data: string]: string} = {};
    public parent: AS.ASObject;
}

export class TemplateRenderer {
    private templates: {[item: string]: TemplateItem};

    public constructor(private templateService: TemplateService, private entityStore: EntityStore) {
    }

    public async prepare() {
        this.templates = await this.templateService.getTemplates();

        for (let templateName in this.templates)
            TemplateRenderer._buildDelegates(this.templates[templateName]);
    }

    private static _buildDelegates(item: TemplateItem) {
        if (item.type == TemplateItemType.Element)
        {
            for (let i in item.arguments)
                for (let subItem of item.arguments[i])
                    TemplateRenderer._buildDelegates(subItem);
            for (let subItem of item.children)
                TemplateRenderer._buildDelegates(subItem);
        }
        else if (item.type == TemplateItemType.Script)
        {
            item.builder = new Function("regs", `let AS = regs.AS; let Renderer = regs.Renderer; let object = regs.object; let item=regs.item; let Data=regs.Data; let parent=regs.parent; return (${item.data});`) as (regs: Registers) => any;
        }
    }

    private _parse(item: TemplateItem, data: AS.ASObject, regs: Registers, override?: boolean): any {
        if (!override && item.type == TemplateItemType.Text) return item.data;

        regs.object = data || {id: null};
        return item.builder(regs);
    }

    private _parseArr(item: TemplateItem[], data: AS.ASObject, regs: Registers): string {
        let result = "";
        for (let it of item)
            result += this._parse(it, data, regs, false);

        return result;
    }

    private _render(item: TemplateItem, data: AS.ASObject, regs: Registers, renderResult: RenderResult, render: boolean, depth: number, element?: HTMLElement, isSub?: boolean): HTMLElement {
        if ("x-render-if" in item.arguments && render) {
            render = this._parse(item.arguments["x-render-if"][0], data, regs, true) as boolean;
        }
        if (depth > 20) { if (element == null) element = document.createElement("span"); element.innerText = "recursion reached"; return element; }
        if (!isSub && ("x-render" in item.arguments || "x-render-id" in item.arguments) && render)
        {
            let itemId = data.id;
            let templateName: string = null;
            if ("x-render" in item.arguments) {
                templateName = item.arguments["x-render"][0].data;
            }
            let template = templateName ? this.templates[templateName] : item;
            let renderId = itemId;
            let oldData = regs.Data;
            let newData: {[name: string]: string} = Object.assign({}, oldData);
            for (let regi in item.arguments) {
                if (regi.startsWith("x-render-"))
                {
                    newData[regi.substr(9)] = this._parseArr(item.arguments[regi], data, regs);
                }
            }

            if ("x-render-id" in item.arguments)
                renderId = this._parse(item.arguments["x-render-id"][0], data, regs, true) as string;

            if (typeof renderId == "object")
            {
                regs.Data = newData;
                let result = this._render(template, renderId as AS.ASObject, regs, renderResult, true, depth + 1, null, true);
                regs.Data = oldData;
                return result;
            }

            if (renderId !== undefined && renderId != itemId)
            {
                let rendered = this._render(template, null, regs, renderResult, false, depth + 1, null, true);
                rendered.dataset["render"] = templateName || "_sub_";
                rendered.dataset["id"] = renderId;
                renderResult.subRender.push({id: renderId, into: rendered, template, data: newData, parent: data});
                return rendered;
            }

            item = template;
            depth += 1;
        }
        if (element == undefined)
            element = document.createElement(item.data);
        else
            while (element.firstChild) element.removeChild(element.firstChild);

        for (let arg in item.arguments) {
            if (!arg.startsWith("x-"))
                element.setAttribute(arg, this._parseArr(item.arguments[arg], data, regs));
        }

        if (!render) return element;

        if ("data-component" in item.arguments) {
            renderResult.componentHandles.push(element);
        }

        if (render) {
            let text = "";
            for (let content of item.children) {
                if (content.type == TemplateItemType.Text || content.type == TemplateItemType.Script) {
                    if (content.type == TemplateItemType.Text)
                        text += content.data;
                    else if (content.type == TemplateItemType.Script)
                        text += this._parse(content, data, regs, false);
                }
                else
                {
                    if (text.length > 0) {
                        element.insertAdjacentHTML("beforeend", text);
                        text = "";
                    }
                    if ("x-for-in" in content.arguments) {
                        let resultItems = this._parse(content.arguments["x-for-in"][0], data, regs, true) as any[];
                        let prevItem = regs.item;
                        for (let subItem of resultItems) {
                            regs.item = subItem;
                            element.appendChild(this._render(content, data, regs, renderResult, true, depth));
                        }
                        regs.item = prevItem;

                        continue;
                    } else if ("x-if" in content.arguments) {
                        let result = this._parse(content.arguments["x-if"][0], data, regs, true) as boolean;
                        if (!result) continue;
                    } else if ("x-else" in content.arguments) {
                        let result = this._parse(content.arguments["x-else"][0], data, regs, true) as boolean;
                        if (result) continue;
                    }

                    element.appendChild(this._render(content, data, regs, renderResult, true, depth));
                }
            }

            if (text.length > 0) {
                element.insertAdjacentHTML("beforeend", text);
                text = "";
            }
        }

        return element;
    }

    public render(template: string|TemplateItem, data: AS.ASObject, elem?: HTMLElement, ndata?: {[name: string]: string}, parent?: AS.ASObject): RenderResult {
        let regs = new Registers();
        regs.AS = new _ASHandler(regs);
        regs.Renderer = new RendererInfo(this.entityStore.session);
        regs.Data = ndata || {};
        regs.parent = parent;
        regs.depth = 0;

        let renderResult = new RenderResult();
        renderResult.subRender = [];
        renderResult.componentHandles = [];
        let result = this._render(typeof(template) === "string" ? this.templates[template] : template, data, regs, renderResult, true, 0, elem, typeof(template) !== "string");
        renderResult.result = result;

        return renderResult;
    }
}

File: Kroeg.Server/Services/Template/TemplateService.cs
using Kroeg.Server.Models;
using Kroeg.Server.Services.EntityStore;
using Kroeg.Server.Tools;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using Jint;
using Kroeg.ActivityStreams;
using HtmlAgilityPack;
using Jint.Parser.Ast;
using System.Text.RegularExpressions;

namespace Kroeg.Server.Services.Template
{
    public class TemplateService
    {
        public Dictionary<string, TemplateItem> Templates { get; } = new Dictionary<string, TemplateItem>();

        private string _base = "templates/";
        private string _baseOverride = "template_override/";

        public string PageTemplate { get; private set; }

        public class Registers {
            public Engine Engine { get; set; }
            public ASHandler Handler { get; set; }
            public RendererData Renderer { get; set; }
            public Dictionary<string, APEntity> UsedEntities { get; set; } = new Dictionary<string, APEntity>();
            public Dictionary<string, string> Data { get; set; } = new Dictionary<string, string>();
            public int EntityCount { get; set; }
        }

        public class ASHandler
        {
            public ASObject obj { get; set; }

            public object[] get(string name)
            {
                if (name == "id")
                    if (obj.Id != null) return new object[] { obj.Id };
                    else return new object[] {};
                if (name == "type") return obj.CompactedTypes.ToArray();
                return obj[name].Select(a => a.Id ?? a.Primitive ?? a.SubObject).ToArray();
            }

            public object take(string name, object def)
            {
                return get(name).FirstOrDefault() ?? def;
            }

            public object take(string name)
            {
                return get(name).FirstOrDefault() ?? "";
            }

            public bool has(string name)
            {
                return get(name).Length > 0;
            }

            public bool contains(string name, object val)
            {
                return get(name).Any(a => a is string && (string) a == (string) val);
            }

            public bool containsAny(string name, object[] vals)
            {
                var check = get(name);
                return vals.Any(a => a is string && check.Any(b => (string) b == (string) a));
            }
        }

        private class doRender {
            public string Template { get; set; }
            public string RenderID { get; set; }
        }

        public class RendererData
        {
            private static Regex _allowed_attributes = new Regex("^(href|rel|class)$");
            private static Regex _allowed_classes = new Regex("^((h|p|u|dt|e)-.*|mention|hashtag|ellipsis|invisible)$");
            private static Regex _disallowed_nodes = new Regex("^(script|object|embed)$");

            public Dictionary<string, string> EmojiContext = new Dictionary<string, string>();

            public string date(string data)
            {
                try {
                    return DateTime.Parse(data).ToLocalTime().ToString();
                } catch (FormatException) { return data; }
            }

            public string sanitize(string data)
            {
                return data.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;");
            }

            private static Regex _emojiRegex = new Regex("([^\\w:]|\n|^):([a-zA-Z0-9_]{2,}):([^\\w:]|$)");

            private void _clean(HtmlNode node)
            {
                if (node.NodeType == HtmlNodeType.Text)
                {
                    var tnode = (HtmlTextNode) node;
                    Match m = Match.Empty;
                    while ((m = _emojiRegex.Match(tnode.Text)).Success)
                    {
                        var first = tnode.Text.Substring(0, m.Index + m.Groups[1].Length);
                        var name = $":{m.Groups[2].Value}:";
                        var second = tnode.Text.Substring(m.Groups[3].Index);

                        tnode.Text = first;
                        if (EmojiContext.ContainsKey(name))
                        {
                            var url = EmojiContext[name];
                            var im = node.OwnerDocument.CreateElement("img");
                            im.Attributes.Add("src", url);
                            im.Attributes.Add("alt", name);
                            im.Attributes.Add("title", name);
                            im.AddClass("emoji");
                            tnode.ParentNode.InsertAfter(im, tnode);
                            
                            tnode = node.OwnerDocument.CreateTextNode();
                            tnode.Text = second;
                            im.ParentNode.InsertAfter(tnode, im);
                        }
                        else
                        {
                            var im = node.OwnerDocument.CreateElement("span");
                            im.AppendChild(node.OwnerDocument.CreateTextNode(name));
                            tnode.ParentNode.InsertAfter(im, tnode);
                            
                            tnode = node.OwnerDocument.CreateTextNode();
                            tnode.Text = second;
                            im.ParentNode.InsertAfter(tnode, im);
                        }
                    }
                }
                if (node.NodeType != HtmlNodeType.Element && node.NodeType != HtmlNodeType.Document) return;
                if (node.NodeType == HtmlNodeType.Element)
                {
                    if (_disallowed_nodes.IsMatch(node.Name.ToLower()))
                    {
                        node.Remove();
                        return;
                    }

                    foreach (var attribute in node.Attributes.ToArray())
                    {
                        if (!_allowed_attributes.IsMatch(attribute.Name)) node.Attributes.Remove(attribute);
                    }

                    if (node.Attributes.Contains("class"))
                    foreach (var cl in node.Attributes["class"].Value.Split(' '))
                    {
                        if (cl.Length > 0 && !_allowed_classes.IsMatch(cl)) node.RemoveClass(cl);
                    }
                }

                foreach (var child in node.ChildNodes.ToArray()) {
                    _clean(child);
                }
            }

            public string clean(string data)
            {
                if (string.IsNullOrWhiteSpace(data)) return "";

                var doc = new HtmlDocument();
                doc.LoadHtml(data);
                _clean(doc.DocumentNode);
                return doc.DocumentNode.InnerHtml;
            }

            public bool server => true;
            public bool client => false;
        }

        public TemplateService()
        {
            Parse();
        }

        public void Parse()
        {
            Templates.Clear();
            _parse(_base);
            _parse(_baseOverride);
        }

        private void _parse(string dir)
        {
            foreach (var file in Directory.EnumerateFiles(dir))
                _parseFile(file);

            foreach (var subdir in Directory.EnumerateDirectories(dir))
                _parse(subdir);
        }

        private void _parseFile(string path)
        {
            if (path == "template_override/.keep") return;

            var data = File.ReadAllText(path);
            if (path == "templates/page.html") PageTemplate = data;
            var templateName = path.Substring(path.IndexOf("/") + 1, path.Length - path.IndexOf("/") - 6).Replace('\\', '/');
            Templates[templateName] = TemplateParser.Parse(data);
        }

        private object _parse(string data, ASObject obj, Registers regs)
        {
            var engine = regs.Engine;
            regs.Handler.obj = obj;

            return regs.Engine.Execute(data).GetCompletionValue().ToObject();
        }

        private async Task<string> _parseElement(HtmlDocument doc, TemplateItem item, IEntityStore entityStore, ASObject data, Registers regs)
        {
            regs.Engine.SetValue("Data", regs.Data);
            regs.Renderer.EmojiContext.Clear();
            var result = doc.CreateElement(item.Data);
            var extraRenderData = new Dictionary<string, string>(regs.Data);
            foreach (var argument in item.Arguments)
            {
                if (!argument.Key.StartsWith("x-") || (argument.Key.StartsWith("x-render-") && argument.Key != "x-render-id"))
                {
                    var resultValue = new StringBuilder();
                    foreach (var subitem in argument.Value)
                        if (subitem.Type == TemplateItemType.Text)
                            resultValue.Append(subitem.Data);
                        else
                            resultValue.Append(_parse(subitem.Data, data, regs));
                    if (argument.Key.StartsWith("x-render-"))
                        extraRenderData[argument.Key.Replace("x-render-", "")] = resultValue.ToString();
                    else
                        result.Attributes.Add(argument.Key, resultValue.ToString());
                }
            }

            bool parse = true;

            if (item.Arguments.ContainsKey("x-if"))
                parse = (bool) _parse(item.Arguments["x-if"][0].Data, data, regs);
            else if (item.Arguments.ContainsKey("x-else"))
                parse = !(bool) _parse(item.Arguments["x-else"][0].Data, data, regs);

            if (!parse) return "";

            if (item.Arguments.ContainsKey("x-render-if"))
                parse = (bool) _parse(item.Arguments["x-render-if"][0].Data, data, regs);

            string err = "";
            
            if ((item.Arguments.ContainsKey("x-render-id") || item.Arguments.ContainsKey("x-render")) && parse)
            {
                var template = item.Arguments.ContainsKey("x-render") ? item.Arguments["x-render"][0].Data : null;
                string id = null;
                ASObject objData = null;
                if (item.Arguments.ContainsKey("x-render-id")) 
                {
                    var res = _parse(item.Arguments["x-render-id"][0].Data, data, regs);
                    if (res is ASObject) objData = (ASObject) res;
                    else id = (string) res;
                }
                else
                    id = data.Id;

                if (objData == null && data.Id == id)
                    objData = data;
                else if (objData == null && id != null)
                {
                    try {

                        APEntity obj = null;
                        if (regs.UsedEntities.ContainsKey(id))
                            obj = regs.UsedEntities[id];
                        else
                            try {
                                obj = await entityStore.GetEntity(id, true);
                            } catch (Exception) { /* nom */ }

                        if (obj != null)
                        {
                            regs.UsedEntities[id] = obj;
                            objData = obj.Data;
                        }

                        regs.EntityCount++;
                    } catch (InvalidOperationException) {
                        err = $"<!-- {id} welp -->";
                    }
                }

                if (objData != null) {
                    var oldData = regs.Data;
                    regs.Data = extraRenderData;
                    string r;
                    if (template != null)
                        r = await _parseTemplate(template, entityStore, objData, regs, doc);
                    else
                        r = await _parseElement(doc, item, entityStore, objData, regs);
                    regs.Data = oldData;
                    return r;
                }
            }
            
            if (item.Arguments.ContainsKey("data-component") && item.Arguments["data-component"][0].Data == "renderhost" && parse)
            {
                var template = item.Arguments["data-template"][0].Data;
                string id = null;
                ASObject objData = null;

                var res = _parse(item.Arguments["data-id"][0].Data, data, regs);
                if (res is ASObject) objData = (ASObject) res;
                else id = (string) res;

                if (objData == null && data.Id == id)
                    objData = data;
                else if (objData == null)
                {
                    APEntity obj;
                    if (regs.UsedEntities.ContainsKey(id))
                        obj = regs.UsedEntities[id];
                    else
                        obj = await entityStore.GetEntity(id, true);

                    if (obj != null)
                    {
                        regs.UsedEntities[id] = obj;
                        objData = obj.Data;
                    }

                    regs.EntityCount++;
                }

                if (objData != null)
                    return await _parseTemplate(template, entityStore, objData, regs, doc);
            }

            if (item.Arguments.ContainsKey("data-component") && item.Arguments["data-component"][0].Data == "emoji" && parse)
            {
                foreach (var tagitem in data["tag"]) {
                    var obj = tagitem.SubObject;
                    if (obj == null)
                    {
                        var entity = await entityStore.GetEntity(tagitem.Id, true);
                        if (entity != null)
                        {
                            regs.UsedEntities[entity.Id] = entity;
                            obj = entity.Data;
                        }
                    }

                    if (obj == null || !obj.Type.Contains("http://joinmastodon.org/ns#Emoji")) continue;

                    var emojiName = (string) obj["name"].First().Primitive;
                    var url = obj["icon"].First().SubObject;

                    if (url == null)
                    {
                        var entity = await entityStore.GetEntity(obj["icon"].First().Id, true);
                        if (entity != null)
                        {
                            regs.UsedEntities[entity.Id] = entity;
                            url = entity.Data;
                        }
                    }

                    if (url == null) continue;

                    regs.Renderer.EmojiContext[emojiName] = url["url"].First().Id;
                }
            }

            var content = new StringBuilder();

            foreach (var subitem in item.Children)
            {
                if (subitem.Type == TemplateItemType.Text)
                    content.Append(subitem.Data);
                else if (subitem.Type == TemplateItemType.Script)
                    content.Append((string) _parse(subitem.Data, data, regs));
                else if (subitem.Type == TemplateItemType.Element)
                {
                    if (subitem.Arguments.ContainsKey("x-for-in"))
                    {
                        var forItems = (object[]) _parse(subitem.Arguments["x-for-in"][0].Data, data, regs);
                        var forIn = "item";
                        foreach (var forItem in forItems)
                        {
                            regs.Engine.SetValue(forIn, forItem);
                            content.Append(await _parseElement(doc, subitem, entityStore, data, regs));
                        }
                    }
                    else
                         content.Append(await _parseElement(doc, subitem, entityStore, data, regs));
                }
            }

            result.InnerHtml = content.ToString() + err;

            return result.OuterHtml;
        }

        private async Task<string> _parseTemplate(string template, IEntityStore entityStore, ASObject data, Registers regs, HtmlDocument doc)
        {

            if (!Templates.ContainsKey(template)) throw new InvalidOperationException($"Template {template} does not exist!");
            var templ = Templates[template];
            return await _parseElement(doc, templ, entityStore, data, regs);
        }

        public async Task<string> ParseTemplate(string template, IEntityStore entityStore, APEntity entity, Registers regs = null)
        {
            if (regs == null) regs = new Registers();
            regs.Handler = new ASHandler();
            regs.Renderer = new RendererData();
            var engine = new Engine();
            engine.SetValue("AS", regs.Handler);
            engine.SetValue("Renderer", regs.Renderer);
            engine.SetValue("Data", regs.Data);
            regs.Engine = engine;
            var doc = new HtmlDocument();

            regs.UsedEntities[entity.Id] = entity;

            var data = await _parseTemplate(template, entityStore, entity.Data, regs, doc);
            Console.WriteLine($"{regs.EntityCount} entities gotten in total, {regs.UsedEntities.Count} separate");
            return data;
        }
    }
}
